src/lib.rs

```diff
@@ -34,21 +40,97 @@ pub fn get_printable_device_list() -> Result<String, Error> {
         let mut device_list = format!("{}\r\n", devices_found);
 
         devices.into_iter().for_each(|device| {
-            let device_info = format!("Id: '{}', Name: '{}'", device.id, device.name);
+            let device_info = format!("Id: '{}', details: '{}'", device.id, device.details);
             info!("{}", device_info);
             device_list = format!("{}\r\n{}", device_list, device_info)
         });
 
         Ok(device_list)
     } else {
-        let no_devices_found = "No device found. Make sure that you connect at least one device with enabled \
-        debug options.";
+        let no_devices_found =
+            "No device found. Make sure that you connect at least one device with enabled \
+             debug options.";
         warn!("{}", no_devices_found);
 
         Ok(String::from(no_devices_found))
     }
 }
 
+pub fn get_printable_app_list(device_id: Option<String>) -> Result<String, Error> {
+    check_too_much_devices(&device_id)?;
+
+    let apps = Backup::list_apps(device_id)?;
+
+    if apps.len() > 0 {
+        let app_found = "Found the following app(s) on device:";
+        info!("{}", app_found);
+
+        let mut app_list = format!("{}\r\n", app_found);
+
+        apps.into_iter().for_each(|app| {
+            let app_name = format!("{}\n", app);
+            info!("{}", app_name);
+            app_list = format!("{}{}", app_list, app_name)
+        });
+
+        Ok(app_list)
+    } else {
+        let no_apps_found = "No packages found.";
+        warn!("{}", no_apps_found);
+
+        Ok(String::from(no_apps_found))
+    }
+}
+
+pub fn backup(
+    device_id: Option<String>,
+    apk: Option<String>,
+    shared: Option<String>,
+    system: Option<String>,
+    only_specified: Option<String>,
+) -> Result<String, Error> {
+    check_too_much_devices(&device_id)?;
+
+    let mut backup_options = backup::BackupOptions::default();
+
+    if let Some(device_id) = device_id {
+        backup_options = backup_options.with_device_id(device_id);
+    }
+    if let Some(_) = apk {
+        backup_options = backup_options.with_applications();
+    }
+    if let Some(_) = shared {
+        backup_options = backup_options.with_shared_storage();
+    }
+    if let Some(_) = system {
+        backup_options = backup_options.with_system_apps();
+    }
+    if let Some(only_specified) = only_specified {
+        backup_options = backup_options.with_only_specified_apps(only_specified);
+    }
+
+    Backup::backup(backup_options)?;
+
+    let backup_finished = "Backup finished.";
+    info!("{}", backup_finished);
+
+    Ok(String::from(backup_finished))
+}
+
+fn check_too_much_devices(device_id: &Option<String>) -> Result<(), Error> {
+    if device_id.is_none() && devices::Device::list_devices()?.len() > 1 {
+        let error =
+            "More than one device connected and no device provided.\n \
+             Please execute adbackup again, with `--device` and one of the following device ids:\n";
```

Comment:
maybe we can execute this ourselves and provide a list here? would be cooler from a usability point of view

Reply:
In the next line the device list is shown to the user with:
``` rust
let error_message = format!("{}\n{}", error, get_printable_device_list()?);
```
So yes, we do that already 😄

Reply:
ah! I thought the `--device` param in the error message was meant to tell the user to do this on his own and not to specify the device, my bad